From b2b649bfc7a739bf22af878e46e38d70e7040c27 Mon Sep 17 00:00:00 2001
From: Steve Best <sbest@redhat.com>
Date: Wed, 27 Aug 2014 19:12:50 -0400
Subject: [ethernet] cxgb4: use remove handler as shutdown handler

Message-id: <20140827191250.2647.29730.sendpatchset@ibm-p7r2-01.lab.bos.redhat.com>
Patchwork-id: 89272
O-Subject: [PATCH RHEL7.1 BZ1123146] net/cxgb4: use remove handler as shutdown handler
Bugzilla: 1123146
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

RHBZ#:
------
https://bugzilla.redhat.com/show_bug.cgi?id=1123146

Description:
------------
Without a shutdown handler, T4 cards behave very badly after a kexec.
Some firmware calls return errors indicating allocation failures, for
example. This is probably because thouse resources were not released by
a BYE message to the firmware, for example.

Using the remove handler guarantees we will use a well tested path.

With this patch I applied, I managed to use kexec multiple times and
probe and iSCSI login worked every time.

Signed-off-by: Thadeu Lima de Souza Cascardo <cascardo@linux.vnet.ibm.com>
Signed-off-by: David S. Miller <davem@davemloft.net>

RHEL Version Found:
-------------------
RHEL 7.0

Brew:
-----
https://brewweb.devel.redhat.com/taskinfo?taskID=7888395

Upstream:
---------
http://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=687d705c031916b83953b714917b04d899e23cf5

Test Status:
------------
Currently being tested by IBM. This is part of their Power KVM product that we are bring into RHEL 7.1.

---------------------------------------------------------------
Steve Best

Proposed Patch:
---------------

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/chelsio/cxgb4/cxgb4_main.c b/drivers/net/ethernet/chelsio/cxgb4/cxgb4_main.c
index 0dd7b16..ab39ae2 100644
--- a/drivers/net/ethernet/chelsio/cxgb4/cxgb4_main.c
+++ b/drivers/net/ethernet/chelsio/cxgb4/cxgb4_main.c
@@ -6179,6 +6179,7 @@ static struct pci_driver cxgb4_driver = {
  .id_table = cxgb4_pci_tbl,
  .probe    = init_one,
  .remove   = remove_one,
+ .shutdown = remove_one,
  .err_handler = &cxgb4_eeh,
 };
 
-- 
1.7.1